pkg/apis/core/v1alpha1/kubernetesdiscovery_types.go

```diff
@@ -18,9 +18,138 @@ limitations under the License.
+type KubernetesDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []KubernetesDiscovery `json:"items"`
+}
+
+// KubernetesDiscoverySpec defines the desired state of KubernetesDiscovery
+type KubernetesDiscoverySpec struct {
+	// Watches determine what resources are discovered.
+	//
+	// If a discovered resource (e.g. Pod) matches the KubernetesWatchRef UID exactly, it will be reported.
+	// If a discovered resource is transitively owned by the KubernetesWatchRef UID, it will be reported.
+	Watches []KubernetesWatchRef `json:"watches"`
+
+	// ExtraSelectors are label selectors that will force discovery of a Pod even if it does not match
+	// the AncestorUID.
+	//
+	// This should only be necessary in the event that a CRD creates Pods but does not set an owner reference
+	// to itself.
+	ExtraSelectors [][]LabelValue `json:"extraSelectors,omitempty"`
+}
+
+// KubernetesWatchRef is similar to v1.ObjectReference from the Kubernetes API and is used to determine
+// what objects should be reported on based on discovery.
+type KubernetesWatchRef struct {
+	// UID is a Kubernetes object UID.
+	//
+	// It should either be the exact object UID or the transitive owner.
+	UID string `json:"uid"`
+	// Namespace is the Kubernetes namespace for discovery. Required.
+	Namespace string `json:"namespace"`
+	// Name is the Kubernetes object name.
+	//
+	// This is not directly used in discovery; it is extra metadata.
+	Name string `json:"name,omitempty"`
+}
+
+// LabelValue is a key-value pair of a Kubernetes label and associated value.
+type LabelValue struct {
+	// Label is the label name.
+	Label string `json:"label"`
+	// Value is the label value.
+	Value string `json:"value"`
+}
+
+var _ resource.Object = &KubernetesDiscovery{}
+var _ resourcestrategy.Validater = &KubernetesDiscovery{}
+
+func (in *KubernetesDiscovery) GetObjectMeta() *metav1.ObjectMeta {
+	return &in.ObjectMeta
+}
+
+func (in *KubernetesDiscovery) NamespaceScoped() bool {
+	return false
+}
+
+func (in *KubernetesDiscovery) New() runtime.Object {
+	return &KubernetesDiscovery{}
+}
+
+func (in *KubernetesDiscovery) NewList() runtime.Object {
+	return &KubernetesDiscoveryList{}
+}
+
+func (in *KubernetesDiscovery) GetGroupVersionResource() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    "tilt.dev",
+		Version:  "v1alpha1",
+		Resource: "kubernetesdiscoveries",
+	}
+}
+
+func (in *KubernetesDiscovery) IsStorageVersion() bool {
+	return true
+}
+
+func (in *KubernetesDiscovery) Validate(ctx context.Context) field.ErrorList {
+	// TODO(user): Modify it, adding your API validation here.
+	return nil
+}
+
+var _ resource.ObjectList = &KubernetesDiscoveryList{}
+
+func (in *KubernetesDiscoveryList) GetListMeta() *metav1.ListMeta {
+	return &in.ListMeta
+}
+
+// KubernetesDiscoveryStatus defines the observed state of KubernetesDiscovery
+type KubernetesDiscoveryStatus struct{}
```

Comment:
Status is temporarily empty to avoid any confusion - see #4478; for sanity + commit/PR size sake, I'm trying to aggressively chunk things and so actual status population/handling is next (first iteration is using the spec to drive discovery, second iteration will actually use status for reporting instead of ye old `PodChangeAction`).

But spoiler: it's just a list of `v1alpha.Pod` which are ⏬ in this file already